evm_loader/program/src/evm/buffer.rs

```diff
@@ -56,131 +48,144 @@ impl Buffer {
             (*ptr).as_ptr()
         };
 
-        Buffer::new(Inner::Account {
+        Buffer::Account {
             key: *account.key,
             data,
             range,
-        })
+        }
     }
 
     #[must_use]
-    pub fn from_vec(v: Vec<u8>) -> Self {
-        if v.is_empty() {
-            return Self::empty();
-        }
-
-        let inner = Inner::Owned { data: v };
-        Self::new(inner)
+    pub fn from_vec(v: Vec<u8>) -> Option<Self> {
+        (!v.is_empty()).then_some(Buffer::Owned(v))
     }
 
     #[must_use]
-    pub fn from_slice(v: &[u8]) -> Self {
+    pub fn from_slice(v: &[u8]) -> Option<Self> {
         Self::from_vec(v.to_vec())
     }
 
     #[must_use]
-    pub fn empty() -> Self {
-        Buffer::new(Inner::Empty)
+    pub fn empty() -> Option<Self> {
+        None
     }
 
     #[must_use]
     pub fn is_initialized(&self) -> bool {
-        !matches!(self.inner, Inner::AccountUninit { .. })
+        !matches!(self, Buffer::AccountUninit { .. })
     }
 
     #[must_use]
     pub fn uninit_data(&self) -> Option<(Pubkey, Range<usize>)> {
-        if let Inner::AccountUninit { key, range } = &self.inner {
+        if let Buffer::AccountUninit { key, range } = self {
             Some((*key, range.clone()))
         } else {
             None
         }
     }
+}
 
-    #[inline]
-    #[must_use]
-    pub fn get_or_default(&self, index: usize) -> u8 {
-        debug_assert!(!self.ptr.is_null());
+#[derive(Clone)]
+pub struct OptionBuffer {
+    // Sacrifice memory to store slice creation components to save discriminating the inner `Buffer`.
+    ptr: *const u8,
+    len: usize,
+    inner: Option<Buffer>,
+}
```

Comment:
Without the optimization, we would be able to get rid of this type. We might still be able to get rid of the `Option` part depending on why we treat empty `Vec`s specially.